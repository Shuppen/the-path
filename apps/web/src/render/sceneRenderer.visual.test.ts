import { describe, expect, it, vi } from 'vitest'

import type { LaneNote, WorldState } from '../world'

const operations: string[] = []

const record = (value: string) => {
  operations.push(value)
}

vi.mock('../environment/reducedMotion', () => ({
  subscribeToReducedMotion: vi.fn((listener: (value: boolean) => void) => {
    listener(false)
    return () => {}
  }),
}))

const createGradient = (kind: 'linear' | 'radial', coords: number[]) => {
  const gradient = {
    addColorStop: vi.fn((offset: number, color: string) => {
      record(`${kind}-stop ${offset.toFixed(2)} ${color}`)
    }),
  }
  record(`${kind}-gradient ${coords.map((value) => value.toFixed(1)).join(',')}`)
  return gradient as unknown as CanvasGradient
}

const createRecordingContext = (): CanvasRenderingContext2D => {
  const canvas = document.createElement('canvas')
  canvas.width = 720
  canvas.height = 1280

  const context: Partial<CanvasRenderingContext2D> = {
    canvas,
    clearRect: (x: number, y: number, w: number, h: number) => record(`clearRect ${x} ${y} ${w} ${h}`),
    createLinearGradient: (...coords: number[]) => createGradient('linear', coords),
    createRadialGradient: (...coords: number[]) => createGradient('radial', coords),
    fillRect: (x: number, y: number, w: number, h: number) =>
      record(`fillRect ${x.toFixed(1)} ${y.toFixed(1)} ${w.toFixed(1)} ${h.toFixed(1)} style=${currentFill}`),
    beginPath: () => record('beginPath'),
    moveTo: (x: number, y: number) => record(`moveTo ${x.toFixed(1)} ${y.toFixed(1)}`),
    lineTo: (x: number, y: number) => record(`lineTo ${x.toFixed(1)} ${y.toFixed(1)}`),
    quadraticCurveTo: (cpx: number, cpy: number, x: number, y: number) =>
      record(
        `quadraticCurveTo ${cpx.toFixed(1)} ${cpy.toFixed(1)} ${x.toFixed(1)} ${y.toFixed(1)} style=${currentFill}`,
      ),
    stroke: () => record(`stroke style=${currentStroke} width=${lineWidth.toFixed(1)}`),
    closePath: () => record('closePath'),
    fill: () => record(`fill style=${currentFill}`),
    save: () => record('save'),
    restore: () => record('restore'),
    arc: (x: number, y: number, radius: number) =>
      record(`arc ${x.toFixed(1)} ${y.toFixed(1)} ${radius.toFixed(1)} style=${currentFill} alpha=${globalAlpha.toFixed(2)}`),
    ellipse: (x: number, y: number, rx: number, ry: number) =>
      record(`ellipse ${x.toFixed(1)} ${y.toFixed(1)} ${rx.toFixed(1)} ${ry.toFixed(1)} style=${currentFill}`),
  }

  let currentFill = '#000'
  let currentStroke = '#000'
  let lineWidth = 1
  let globalAlpha = 1

  Object.defineProperty(context, 'fillStyle', {
    get: () => currentFill,
    set: (value) => {
      currentFill = String(value)
      record(`set fillStyle ${currentFill}`)
    },
  })

  Object.defineProperty(context, 'strokeStyle', {
    get: () => currentStroke,
    set: (value) => {
      currentStroke = String(value)
      record(`set strokeStyle ${currentStroke}`)
    },
  })

  Object.defineProperty(context, 'lineWidth', {
    get: () => lineWidth,
    set: (value) => {
      lineWidth = value
      record(`set lineWidth ${lineWidth.toFixed(1)}`)
    },
  })

  Object.defineProperty(context, 'globalAlpha', {
    get: () => globalAlpha,
    set: (value) => {
      globalAlpha = value
      record(`set globalAlpha ${globalAlpha.toFixed(2)}`)
    },
  })

  return context as CanvasRenderingContext2D
}

const laneWidth = (720 - 48) / 4
const hitLineY = 1280 * (1 - 0.12 * 0.5)

let noteId = 0

const createNote = (overrides: Partial<LaneNote>): LaneNote => ({
  id: overrides.id ?? (noteId += 1),
  lane: overrides.lane ?? 0,
  time: overrides.time ?? 0,
  judged: overrides.judged ?? false,
  judgement: overrides.judgement,
  hitTime: overrides.hitTime,
})

const createState = (): WorldState => {
  // ensure deterministic ids for each state instantiation
  noteId = 0
  return {
    seed: 'visual',
    time: 42.4,
    beat: 120,
    status: 'running',
    stage: {
      width: 720,
      height: 1280,
      hitLineY,
      laneWidth,
      lanePadding: 24,
      laneCount: 4,
      scrollSpeed: 720,
    },
    lanes: { count: 4 },
    notes: [
      createNote({ id: 1, lane: 0, time: 42.65 }),
      createNote({ id: 2, lane: 1, time: 42.36, judged: true, judgement: 'perfect', hitTime: 42.36 }),
      createNote({ id: 3, lane: 2, time: 41.98 }),
    ],
    runner: {
      lane: 1,
      targetLane: 2,
      transitionFrom: 1,
      transitionStart: 42.2,
      transitionDuration: 0.15,
      combo: 18,
      bestCombo: 30,
      score: 98200,
      perfectHits: 40,
      goodHits: 6,
      missHits: 2,
    },
    feedback: [
      { id: 1, judgement: 'perfect', createdAt: 42.1, x: 24 + laneWidth * 1.5, y: hitLineY },
      { id: 2, judgement: 'miss', createdAt: 41.9, x: 24 + laneWidth * 2.5, y: hitLineY },
    ],
    accuracy: 0.94,
  }
}

const { SceneRenderer } = await import('./sceneRenderer')

describe('SceneRenderer visual output', () => {
  it('records drawing operations for snapshot comparison', () => {
    operations.length = 0
    const ctx = createRecordingContext()
    const renderer = new SceneRenderer(ctx)

    renderer.render(createState())

    expect(operations).toMatchSnapshot()
  })
})
