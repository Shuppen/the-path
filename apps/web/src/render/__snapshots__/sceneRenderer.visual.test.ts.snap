// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SceneRenderer textures > applies texture fills and sprite drawing when assets are available 1`] = `
[
  "clearRect 0 0 800 600",
  "save",
  "set fillStyle pattern:background",
  "fillRect 0 0 800 600 style=pattern:background",
  "restore",
  "set fillStyle gradient:radial(400.00,300.00,60.00,400.00,300.00,480.00)[0.00:rgba(14, 165, 233, 0.135)|0.25:rgba(13, 148, 136, 0.09)|1.00:#020617]",
  "fillRect 0 0 800 600 style=gradient:radial(400.00,300.00,60.00,400.00,300.00,480.00)[0.00:rgba(14, 165, 233, 0.135)|0.25:rgba(13, 148, 136, 0.09)|1.00:#020617]",
  "save",
  "set fillStyle pattern:ground",
  "fillRect 0 500 800 100 style=pattern:ground",
  "restore",
  "set fillStyle gradient:linear(0.00,500.00,0.00,600.00)[0.00:rgba(15, 118, 110, 0.45)|1.00:rgba(15, 23, 42, 0.7)]",
  "fillRect 0 500 800 100 style=gradient:linear(0.00,500.00,0.00,600.00)[0.00:rgba(15, 118, 110, 0.45)|1.00:rgba(15, 23, 42, 0.7)]",
  "set strokeStyle rgba(56, 189, 248, 0.2)",
  "set lineWidth 2",
  "beginPath",
  "moveTo 0 500",
  "lineTo 800 500",
  "stroke style=rgba(56, 189, 248, 0.2) width=2",
  "save",
  "set fillStyle pattern:obstacle-spire",
  "beginPath",
  "moveTo 265 340",
  "lineTo 290 460",
  "lineTo 240 460",
  "closePath",
  "fill style=pattern:obstacle-spire",
  "restore",
  "set strokeStyle rgba(244, 244, 255, 0.2)",
  "set lineWidth 1",
  "stroke style=rgba(244, 244, 255, 0.2) width=1",
  "save",
  "set fillStyle pattern:obstacle-pulse",
  "beginPath",
  "roundRect 360 360 60 90 10",
  "fill style=pattern:obstacle-pulse",
  "restore",
  "save",
  "set fillStyle pattern:obstacle-block",
  "beginPath",
  "roundRect 500 320 80 140 10",
  "fill style=pattern:obstacle-block",
  "restore",
  "set strokeStyle rgba(244, 244, 255, 0.2)",
  "set lineWidth 1",
  "stroke style=rgba(244, 244, 255, 0.2) width=1",
  "save",
  "translate 0 0",
  "drawImage 120 380 48 72",
  "set globalAlpha 0.45",
  "set fillStyle rgba(56, 189, 248, 0.45)",
  "beginPath",
  "fill style=rgba(56, 189, 248, 0.45)",
  "restore",
]
`;
